File: Sort/BubbleSort.c
#include<stdio.h>
#include <time.h>
#include <stdlib.h>
#include <math.h>

#define bool int
#define false 0
#define true 1


/* 优化的冒泡算法 */
/*
* 优化方法:对原本有序的数组,不进行排序
* */
void Print(int *,int);


void BubbleSort(int *a,unsigned length)
{
    printf("\n排序前:\n");
    Print(a,length);
    bool flag = true;
    int i ;
    int j;
    int temp;
    int count = 0; //比较的次数

    for(i = 0; i < length -1 && flag ;i++){
        flag = false;
        for(j = i+1;j < length ;j++){
            if(a[j] < a[i]){
                temp = a[i];
                a[i] = a[j];
                a[j] = temp;
                count++;
                flag = true;  
            }
        }
    }
    printf("\n\n排序后:");
    Print(a,length);
    printf("\n比较的次数:%d\n",count);

}

void Print(int *a,int length)
{
    int i = 0;
    for(i = 0 ;i < length;i++)
    {
        if(i%10 == 0)
            putchar(10);
        printf("%-4d",a[i]);
    }
}

int main()
{
    int a[100] = {0};
    for(int i = 0;i < 100;i++)
    {
        srand(i*time(NULL));
        a[i] = rand()%500;
    }
    BubbleSort(a,100);
    
}

File: sort_go/bubble_sort.go
package main

import "fmt"

// BubbleSort 冒泡排序
// 时间复杂度:
//
//	 特性: 稳定的排序, 优化后最快o(n),即对有序数组排序
//		1.未优化时: O(n²)
//			外循环n次 = n
//			内循环 n-1,n-2,n-3...1 次 = (n-1)/2
//			总: (n-1)/2 * n ≈ n²
//	 2.引入flag标记优化: o(n) 至 o(n²)
func BubbleSort(arr []int) {
	flag := 0
	for i := 0; i < len(arr); i++ {
		for j := 1; j < len(arr)-i; j++ {
			if arr[j-1] > arr[j] {
				arr[j-1], arr[j] = arr[j], arr[j-1]
				flag = 1
			}
		}
		if flag == 0 {
			break
		}
	}
}

func main() {
	arr := []int{7, 1, 4, 6, 9, 2, 5, 3}
	BubbleSort(arr)
	fmt.Println(arr)

	arr = []int{1, 2, 1, 1, 1, 1, 2}
	BubbleSort(arr)
	fmt.Println(arr)

	arr = []int{}
	BubbleSort(arr)
	fmt.Println(arr)
}
